chore: add example how to use data loaders with suspendable functions (#1678)
ExpediaGroup/graphql-kotlin: website/docs/server/data-loader/data-loader.md

}
```

## DataLoaders and Coroutines

> `graphql-java` relies on `CompletableFuture`s for scheduling and asynchronously execute GraphQL operations.

Reviewer: ```suggestion
`graphql-java` relies on `CompletableFuture`s for scheduling and asynchronously executing GraphQL operations.
```

While we can provide native support for coroutines for data fetchers (aka field resolvers) because they are resolved
independently, we cannot easily provide native support for the `DataLoader` pattern as it relies

> on `CompletableFuture` and we cannot update it to use coroutines state machine logic without fully rewriting
> GraphQL Java execution engine.

Reviewer: ```suggestion
on `CompletableFuture` state machine internals and we cannot update it to use coroutines without fully rewriting
GraphQL Java execution engine.
```

> If you would like to use `DataLoader` pattern in your project, you have to update your data fetchers (aka field resolvers) to return
> `CompletableFuture` from the invoked `DataLoader` instead of marking it a `suspend`.

Reviewer: ```suggestion
If you would like to use `DataLoader` pattern in your project, you have to update your data fetchers (aka field resolvers) to return
`CompletableFuture` from the invoked `DataLoader`.
```

}
```

## DataLoaders and Coroutines

`graphql-java` relies on `CompletableFuture` for scheduling and execute asynchronous field resolvers (aka data fetchers),
`graphql-java` deliberately aims for near zero dependencies which means no Kotlin / WebFlux / RxJava.

Similar to `asynchronous` field resolvers (aka data fetcher), `DataLoader` pattern implementation in `graphql-java` works with `CompletableFuture`,

> and because of the listed `graphql-java` constrains we don't provide any native support for `DataLoader` pattern using suspendable functions. Instead, return

Reviewer: IMHO previous version was clearer - I don't think adding the section about the data fetchers is needed and it brings confusion, i.e. yes `graphql-java` does use CF for data fetchers but we provide interop with coroutines but also `graphql-java` uses CF for data loaders but we don't provide support for it?

Adding details on why it works on data fetchers but not on data loaders might be too much so I'd just keep the previous comment.

```kotlin
class MyQuery(
    private val userService: UserService
) : Query {
    suspend fun getUser(id: Int): User = userService.getUser(id)
}

class UserService {
    suspend fun getUser(id: Int): User = // async logic to get user
    suspend fun getUsers(ids: List<Int>): List<User> = // async logic to get users
}
```

When the execution of the query completes we will have a total of 2 requests to the `UserService`, this is where the usage of
`DataLoader` pattern can help to make the execution of query more performant by making only 1 request.

Lets create the `UserDataLoader`:

```kotlin
class UserDataLoader : KotlinDataLoader<ID, User> {
    override val dataLoaderName = "UserDataLoader"
    override fun getDataLoader() =
        DataLoaderFactory.newDataLoader<Int, User> { ids, batchLoaderEnvironment ->
            val coroutineScope =
                batchLoaderEnvironment.getGraphQLContext()?.get<CoroutineScope>()
                    ?: CoroutineScope(EmptyCoroutineContext)

            coroutineScope.future {
                userService.getUsers(ids)
            }
        }
}
```

There are some things going on here:

> 1. We define the `UserDataLoader` with name "UserDataLoader"

Reviewer: Maybe add comments to the code above to indicate what are you referring to?

i.e. something along those lines
```kotlin
class UserDataLoader : KotlinDataLoader<ID, User> { // 1
```

> 2. The `getLoader()` method returns a `DataLoader<Int, User>`, which `BatchLoader` function should return a `List<User>`

Reviewer: there is no `getLoader()` method, did you mean `getDataLoader()`?

> 3. Given that we **don't want** to change our `UserService` async model that is using coroutines, we need a `CoroutineScope`, [which is conveniently available](../../schema-generator/execution/async-models/#coroutines) in the `GraphQLContext`, which is provided in the `DataFetchingEnvironment` and accessible with the [getGraphQLContext()](https://github.com/ExpediaGroup/graphql-kotlin/blob/master/executions/graphql-kotlin-dataloader-instrumentation/src/main/kotlin/com/expediagroup/graphql/dataloader/instrumentation/extensions/BatchLoaderEnvironmentExtensions.kt#L43) extension function.

Reviewer: ```suggestion
3. Given that we **don't want** to change our `UserService` async model that is using coroutines, we need a `CoroutineScope`, [which is conveniently available](../../schema-generator/execution/async-models/#coroutines) in the `GraphQLContext` and accessible through [`DataFetchingEnvironment#getGraphQLContext()`](https://github.com/ExpediaGroup/graphql-kotlin/blob/master/executions/graphql-kotlin-dataloader-instrumentation/src/main/kotlin/com/expediagroup/graphql/dataloader/instrumentation/extensions/BatchLoaderEnvironmentExtensions.kt#L43) extension function.
```

4. After retrieving the `CoroutineScope` from the `batchLoaderEnvironment` we will be able to execute the `userService.getUsers(ids)` suspendable function.
5. We interoperate the suspendable function result to a `CompletableFuture` using [coroutineScope.future](https://kotlinlang.org/api/kotlinx.coroutines/kotlinx-coroutines-jdk8/kotlinx.coroutines.future/future.html)

> Finally, the only thing that we need to change is the `user` field resolver, to not be suspendable and
> just return the `CompletableFuture<User>` that the `DataLoader`, make sure to pass the `dataFetchingEnvironment` as `keyContext` which is the second argument of `DataLoader.load`

Reviewer: ```suggestion
Finally, we need to update `user` field resolver, to return the `CompletableFuture<User>` from the invoked `DataLoader`. Make sure to update method signature to also accept the `dataFetchingEnvironment` as you need to pass it to `DataLoader#load` method to be able to execute the request in appropriate coroutine scope.
```
